src/lib/logical_query_plan/abstract_lqp_node.cpp

```diff
@@ -241,13 +242,62 @@ ColumnID AbstractLQPNode::get_column_id(const AbstractExpression& expression) co
   return *column_id;
 }
 
+std::optional<const std::shared_ptr<LQPColumnExpression>> AbstractLQPNode::find_column_expression(
+    const ColumnID column_id) const {
+  const auto& output_expressions = this->output_expressions();
+
+  for (const auto& output_expression : output_expressions) {
+    const auto column_expression = dynamic_pointer_cast<LQPColumnExpression>(output_expression);
+    if (column_expression && column_expression->original_column_id == column_id) {
+      return column_expression;
+    }
+  }
+  return std::nullopt;
+}
+
 bool AbstractLQPNode::is_column_nullable(const ColumnID column_id) const {
   // Default behaviour: Forward from input
   Assert(left_input() && !right_input(),
          "Can forward nullability from input iff there is a left input and no right input");
   return left_input()->is_column_nullable(column_id);
 }
 
+std::shared_ptr<LQPUniqueConstraints> AbstractLQPNode::unique_constraints() const {
+  Fail("Node misses implementation for unique constraints.");
+}
+
+bool AbstractLQPNode::has_unique_constraint(const ExpressionUnorderedSet& output_expressions_subset) const {
+  if constexpr (HYRISE_DEBUG) {
+    Assert(!output_expressions_subset.empty(), "An empty input expressions set does not make sense.");
+
+    const auto& output_expressions = this->output_expressions();
+    Assert(std::all_of(output_expressions_subset.cbegin(), output_expressions_subset.cend(),
+                       [&output_expressions](const auto& input_expression) {
+                         return std::any_of(output_expressions.cbegin(), output_expressions.cend(),
+                                            [&input_expression](const auto& output_expression) {
+                                              return *output_expression == *input_expression;
+                                            });
+                       }),
+           "The given expressions are not a subset of the LQP's output expressions.");
+  }
+
+  const auto& unique_constraints = this->unique_constraints();
+  if (unique_constraints->empty()) return false;
+
+  // Look for a unique constraint that is solely based on the given expressions
+  for (const auto& unique_constraint : *unique_constraints) {
+    if (unique_constraint.expressions.size() == output_expressions_subset.size() &&
+        std::all_of(
+            output_expressions_subset.cbegin(), output_expressions_subset.cend(),
+            [&](const auto output_expression) { return unique_constraint.expressions.contains(output_expression); })) {
+      // Found match
+      return true;
+    }
+  }
+  // No match
+  return false;
+}
+
 std::vector<FunctionalDependency> AbstractLQPNode::functional_dependencies() const {
```

Comment:
At which point are unique constraints created, e.g., by an AggregateNode reflected in the FDs?

Reply:
Unique constraints are created when `unique_constraints()` is called.

>  e.g., by an AggregateNode reflected in the FDs?

Can you explain this a bit further?

Reply:
In StoredTableNode, we create FDs based on non-nullable unique columns. With the AggregateNode, we introduce new unique columns. Wouldn't this result in new FDs?

Reply:
It would, but is not yet implemented. I could think of two solutions:
* Override `functional_dependencies()` in AggregateNode, creating an edge case.
* Add some additional logic to AbstractLQPNode for creating FDs from unique constraints. 

I prefer the last option, which could be work for another PR.

Reply:
Hmm, I think we need to make functional_dependencies() virtual to deal with joins and unions anyway. Right now, it assumes that FDs from both the left and the right persist in the output. That is true for joins, but not for unions. We should add an assertion immediately to prevent this broken behavior from being called.

Speaking about joins, it also looks like the FDs introduced by 1:1/1:N joins are not covered yet. That would be another argument for making it virtual.

However, it only make sense to go down this rabbit hole if we plan to use it more than we do right now. Let's discuss this on Wednesday.

Reply:
I added 

```cpp
  Assert(type != LQPNodeType::Union || fds_left != fds_right,
         "We do not yet support FD-forwarding for UnionNodes with differing input FDs.");
```

As discussed, I will address the issue in a separate PR.

Reply:
I already started work on making `functional_dependencies()` pure virtual.  I noticed that UnionNode already overrides the default implementation:

```cpp
std::vector<FunctionalDependency> UnionNode::functional_dependencies() const {
  switch (set_operation_mode) {
    case SetOperationMode::Unique:
    case SetOperationMode::All:
      Fail("Handling of functional dependencies is not yet specified for UNION (ALL)");
    case SetOperationMode::Positions:
      // By definition, UnionPositions requires both input tables to have the same table origin and structure.
      // Therefore, we can forward the FDs of either the left or right input node.
      DebugAssert(left_input()->functional_dependencies() == right_input()->functional_dependencies(),
                  "Expected both input nodes to have the same FDs.");
      return left_input()->functional_dependencies();
  }
  Fail("Unhandled UnionMode");
}
```

Therefore, I removed the Assert from above.